index.bs

```diff
@@ -898,57 +910,32 @@ partial interface MLContext {
-            1. [=reject=] |promise| with an "{{OperationError}}" {{DOMException}} and stop.
-        1. Else:
-            1. Let |outputTensor| be the output tensor returned by |graph|.{{MLGraph/[[implementation]]}}.
-            1. Let |outputDesc| be |graph|.{{MLGraph/[[outputDescriptors]]}}[|key|].
-            1. If the data type of |outputTensor| doesn't match the [=element type=] of {{ArrayBufferView}} |value|, then:
-                1. [=reject=] |promise| with a "{{DataError}}" {{DOMException}} and stop.
-            1. If the byte length of |outputTensor| is not equal to [=byte length=] of |outputDesc|, then:
-                1. [=reject=] |promise| with an "{{OperationError}}" {{DOMException}} and stop.
-            1. Else:
-                1. Set the values of |value| to the values of |outputTensor|.
-            1. If all compute requests are completed, then:
-                1. Let |result| be a new {{MLComputeResult}}.
-                1. Set |result|.{{MLComputeResult/inputs}} to |transferredInputs|.
-                1. Set |result|.{{MLComputeResult/outputs}} to |transferredOutputs|.
-                1. [=Resolve=] |promise| with |result| and stop.
-    1. Return |promise|.
+    1. Return |promise| and run the following steps [=in parallel=]:
+        1. If |graph|.{{MLGraph/[[context]]}}.{{MLContext/[[contextType]]}} is not "[=default-context|default=], [=reject=] |promise| with an "{{OperationError}}" {{DOMException}} and stop.
+        1. If invoking the <a>validate graph resources</a> algorithm given |inputs| and |graph|.{{MLGraph/[[inputDescriptors]]}} returns `false`, then [=reject=] |promise| with a "{{DataError}}" {{DOMException}} and stop.
+        1. If invoking the <a>validate graph resources</a> algorithm given |outputs| and |graph|.{{MLGraph/[[outputDescriptors]]}} returns `false`, then [=reject=] |promise| with a "{{DataError}}" {{DOMException}} and stop.
+        1. Let |transferredInputs| be the result of [=MLNamedArrayBufferViews/transfer|transferring=] {{MLNamedArrayBufferViews}} |inputs|.
+        1. Let |transferredOutputs| be the result of [=MLNamedArrayBufferViews/transfer|transferring=] {{MLNamedArrayBufferViews}} |outputs|.
+        1. Invoke <a>execute graph</a> given |graph|, |transferredInputs| and |transferredOutputs|.
+        1. If that throws an error, [=reject=] |promise| with the error and stop.
+        1. Otherwise, when <a>execute graph</a> has completed:
+            1. Let |result| be a new {{MLComputeResult}}.
+            1. Set |result|.{{MLComputeResult/inputs}} to |transferredInputs|.
+            1. Set |result|.{{MLComputeResult/outputs}} to |transferredOutputs|.
+            1. [=Resolve=] |promise| with |result| and stop.
+
+    <div class="note">
+        During asynchronous execution, the compute requests in the <a>execute graph</a> algorithm MAY be executed [=in parallel=], or otherwise optimized for the underlying platform.
+    </div>
```

Comment:
I think this paragraph may add more confusion to the implementation than it helps. In practice, multiple asynchronous executions on the CPU may be handled independently by multiple worker threads. But for the GPU, although the recording of multiple command lists can be done concurrently by multiple CPU worker threads, the execution of the command lists still happen sequentially (but on the GPU timeline) once they are submitted to the command queue.

Reply:
Right, that's an important distinction, from the perspective of the underlying platform. The (informative) note was written in the web platform terminology and from web spec viewpoint. I wonder should we extend the note with your remark? Then it would be informative AND exact. But I am also fine removing it.

@anssiko, @dontcallmedom do you have an input from the web perspective, is this kind of note useful?
@wchao1115 and @huningxin,  what would you suggest: remove, extend, or rewrite (please suggest text)?

Reply:
> In practice, multiple asynchronous executions on the CPU may be handled independently by multiple worker threads.

For CPU execution, if multiple worker threads execute the graph "in parallel", it would require synchronization between them (e.g. a lock) that protects the graph data structure and also ensures the execution order. It increases the implementation complexity and probably would be easier to introduce bugs. 

Because Chromium also prefers to [Using Sequences Instead of Locks](https://chromium.googlesource.com/chromium/src/+/main/docs/threading_and_tasks.md#Using-Sequences-Instead-of-Locks), this is a proposal [CL](https://chromium-review.googlesource.com/c/chromium/src/+/4382891) that uses `SequencedTaskRunner` of worker pool for XNNPACK CPU backend. With that, although there may be multiple worker pool threads are scheduled to compute the graph, `SequencedTaskRunner` would ensure there is one compute at a time in the posting order. 

/cc @wacky6.

Reply:
> In practice, multiple asynchronous executions on the CPU may be handled independently by multiple worker threads.

I think this is unlikely to be implemented in practice:

- Backends (most likely) aren't re-entrant safe. The implementation need to clone them (e.g. XNN runtime and static values) for each thread for safety. Resource usage is `O(N_threads)`, which might be too much for large models.
- Browsers likely wants to throttle execution (e.g. for background pages) or limit resource usage or preempt under pressure (e.g. preventing an website to consume all available RAM). Single executor (task runner) can achieve this much easier.
- Single executor likely incurs less overhead (e.g. less resource contention and context switches when n_threads > n_cores).
- Accelerators (GPU, TPU, NPU) probably follows the sequenced execution order.

Reply:
Note the expression ["in parallel"](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel) in web specs mainly means it should not block the event loop. Details on how that "parallelism" is interpreted are not needed to be explicit. Quoting from the HTML spec,

> To run steps in parallel means those steps are to be run, one after another, at the same time as other logic in the standard (e.g., at the same time as the [event loop](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop)). This standard does not define the precise mechanism by which this is achieved, be it time-sharing cooperative multitasking, fibers, threads, processes, using different hyperthreads, cores, CPUs, machines, etc. By contrast, an operation that is to run immediately must interrupt the currently running task, run itself, and then resume the previously running task.

The [guidance ](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors) for spec authors is:

> the overriding guideline is that any work a specification needs to perform that would otherwise block the [event loop](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop) must instead be performed [in parallel](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel) with it.  This includes (but is not limited to): performing heavy computation; displaying a user-facing prompt; performing operations which could require involving outside systems (i.e. "going out of process").

Reply:
FWIW I think we can delete this note, since the normative text already contains how to run the steps.

Reply:
@wchao1115 , @huningxin : based on the content of this discussion, are you fine keeping this note, or should it be removed?

Reply:
Per https://www.w3.org/TR/webnn/#w3c-conventions RFC 2119 terminology is not to be used in an informative note.

My recommendation would be to remove this note.